query_optimizer/ExecutionGenerator.cpp

```diff
@@ -1020,6 +1024,90 @@ void ExecutionGenerator::convertInsertTuple(
                                        false /* is_pipeline_breaker */);
 }
 
+void ExecutionGenerator::convertInsertSelection(
+    const P::InsertSelectionPtr &physical_plan) {
+  // InsertSelection is converted to a Select and a SaveBlocks.
+
+  const CatalogRelationInfo *destination_relation_info =
+      findRelationInfoOutputByPhysical(physical_plan->destination());
+  const CatalogRelation &destination_relation =
+      *optimizer_context_->catalog_database()->getRelationByIdMutable(
+          destination_relation_info->relation->getID());
+
+  // FIXME(qzeng): A better way is using a traits struct to look up whether a storage
+  //               block supports ad-hoc insertion instead of hard-coding the block types.
+  const StorageBlockLayout &storage_block_layout =
+      destination_relation.getDefaultStorageBlockLayout();
+  if (storage_block_layout.getDescription().tuple_store_description().sub_block_type() ==
+          TupleStorageSubBlockDescription::COMPRESSED_COLUMN_STORE
+      || storage_block_layout.getDescription().tuple_store_description().sub_block_type() ==
+             TupleStorageSubBlockDescription::COMPRESSED_PACKED_ROW_STORE) {
+    THROW_SQL_ERROR() << "INSERT statement is not supported for the relation "
+                      << destination_relation.getName()
+                      << ", because its storage blocks do not support ad-hoc insertion";
+  }
+
+  // Create InsertDestination proto.
+  const QueryContext::insert_destination_id insert_destination_index =
+      query_context_proto_->insert_destinations_size();
+  S::InsertDestination *insert_destination_proto = query_context_proto_->add_insert_destinations();
+
+  insert_destination_proto->set_insert_destination_type(S::InsertDestinationType::BLOCK_POOL);
+  insert_destination_proto->set_relation_id(destination_relation.getID());
+
+  const vector<block_id> blocks(destination_relation.getBlocksSnapshot());
+  for (const block_id block : blocks) {
+    insert_destination_proto->AddExtension(S::BlockPoolInsertDestination::blocks, block);
+  }
+
+  const CatalogRelationInfo *selection_relation_info =
+      findRelationInfoOutputByPhysical(physical_plan->selection());
+  CatalogRelation *selection_relation =
+      optimizer_context_->catalog_database()->getRelationByIdMutable(
+          selection_relation_info->relation->getID());
```

Comment:
Delete this one, and replace `selection_relation` in Line 1087 with `*selection_relation_info->relation`.